File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Model/Factory/ProductFactory.cs
using System.Collections.Generic;
using Bogus;
using RekindleTaf.Exceptions;
using RekindleTaf.Utils;

namespace RekindleTaf.Model.Factory;

public class ProductFactory : IFactory<Product>
{
    private static readonly Faker<Product> ActiveProductFaker = new Faker<Product>()
        .RuleFor(product => product.Name, (faker) => faker.Commerce.Product())
        .RuleFor(product => product.Available, () => true)
        .RuleFor(product => product.Price, () => Randomizers.RandomPriceLessThanThousand());

    private static readonly Faker<Product> InActiveProductFaker = new Faker<Product>()
        .RuleFor(product => product.Name, (faker) => faker.Commerce.Product())
        .RuleFor(product => product.Available, () => false)
        .RuleFor(product => product.Price, () => Randomizers.RandomPriceLessThanThousand());

    private IDictionary<IValueKey, Supplier<Product>> Map = new Dictionary<IValueKey, Supplier<Product>>()
    {
        {ProductKey.StdProductAvailable, () => ActiveProductFaker.Generate()},
        {ProductKey.StdProductNonAvailable, () => InActiveProductFaker.Generate()}
    };

    public Product Create(IValueKey key)
    {
        if (Map.TryGetValue(key, out Supplier<Product>? value))
        {
            return value.Invoke();
        }

        throw new ValueObjectNotFoundException($"Product identified by {key} has not yet been implemented");
    }
}

File: typescript-project/rekindle-book-store-aqa/src/model/product-factory.ts
import {Product} from "@/model/domain-model";
import {faker} from '@faker-js/faker';
import {Factory} from "@/model/factory";
import {randomPriceLessThanThousand} from "../utils/utils";

export const ProductKey = {
    stdProductActive: Symbol('StdProductActive'),
    stdProductInactive: Symbol('StdProductInactive')
};

export class ProductFactory implements Factory<Product> {

    private readonly map: Map<Symbol, Supplier<Product>> = new Map([
        [ProductKey.stdProductActive, () => this.createActiveProduct()],
        [ProductKey.stdProductInactive, () => this.createInActiveProduct()]
    ]);

    private createActiveProduct(): Product {
        return {
            name: faker.commerce.productName(),
            available: true,
            price: randomPriceLessThanThousand()
        }
    }

    private createInActiveProduct(): Product {
        return {
            name: faker.commerce.productName(),
            available: false,
            price: randomPriceLessThanThousand()
        }
    }

    create(key: Symbol): Product {
        const supplier = this.map.get(key);
        if (supplier) {
            return supplier();
        }
        throw new Error(`Bookstore identified by ${key.toString()} has not yet been implemented`);
    }

}